Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp

```diff
@@ -0,0 +1,92 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Detector/IndexedRootVolumeFinderBuilder.hpp"
+
+#include "Acts/Detector/DetectorVolume.hpp"
+#include "Acts/Detector/detail/CylindricalDetectorHelper.hpp"
+#include "Acts/Detector/detail/GridAxisGenerators.hpp"
+#include "Acts/Navigation/DetectorVolumeFinders.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+
+namespace {
+
+template <typename Grid2D>
+void fillGridIndices2D(
+    const Acts::GeometryContext& gctx, Grid2D& grid,
+    const std::vector<std::shared_ptr<Acts::Experimental::DetectorVolume>>&
+        rootVolumes,
+    const std::array<std::vector<Acts::ActsScalar>, 2u>& boundaries,
+    const std::array<Acts::BinningValue, 2u>& casts) {
+  // Brute force loop over all bins & all volumes
+  for (const auto [ic0, c0] : Acts::enumerate(boundaries[0u])) {
+    if (ic0 > 0) {
+      Acts::ActsScalar v0 = 0.5 * (c0 + boundaries[0u][ic0 - 1]);
+      for (const auto [ic1, c1] : Acts::enumerate(boundaries[1u])) {
+        if (ic1 > 0) {
+          Acts::ActsScalar v1 = 0.5 * (c1 + boundaries[1u][ic1 - 1]);
+          if (casts ==
```

Comment:
Can't you do that check once at the beginning, outside of the loop?

Reply:
For now, yes, but I am planing to change the casting into compile time switches, then this will change anyway.